llama_index/query_engine/knowledge_graph_query_engine.py

```diff
@@ -0,0 +1,178 @@
+< RETURN p.name;
+---
+> MATCH (p:`person`)-[:directed]->(m:`movie`) WHERE m.`movie`.`name` == 'The Godfather'
+> RETURN p.`person`.`name`;
+```
+
+Question: {query_str}
+
+NebulaGraph Cypher dialect query:
+"""
+DEFAULT_NEBULAGRAPH_NL2CYPHER_PROMPT = Prompt(
+    DEFAULT_NEBULAGRAPH_NL2CYPHER_PROMPT_TMPL,
+    prompt_type=PromptType.TEXT_TO_GRAPH_QUERY,
+)
+
+DEFAULT_NL2GRAPH_PROMPT_MAP = {
+    GraphStoreType.NEBULA: DEFAULT_NEBULAGRAPH_NL2CYPHER_PROMPT,
+}
+
+DEFAULT_KG_RESPONSE_ANSWER_PROMPT_TMPL = """
+The original question is given below.
+This question has been translated into a Graph Database query.
+Both the Graph query and the response are given below.
+Given the Graph Query response, synthesise a response to the original question.
+
+Original question: {query_str}
+Graph query: {kg_query_str}
+Graph response: {kg_response_str}
+Response: 
+"""
+
+DEFAULT_KG_RESPONSE_ANSWER_PROMPT = Prompt(
+    DEFAULT_KG_RESPONSE_ANSWER_PROMPT_TMPL,
+    prompt_type=PromptType.QUESTION_ANSWER,
+)
+
+
+class KnowledgeGraphQueryEngine(BaseQueryEngine):
+    """Knowledge graph query engine.
+
+    Query engine to call a knowledge graph.
+
+    Args:
+        service_context (Optional[ServiceContext]): A service context to use.
+        storage_context (Optional[StorageContext]): A storage context to use.
+        refresh_schema (bool): Whether to refresh the schema.
+        verbose (bool): Whether to print intermediate results.
+        **kwargs: Additional keyword arguments.
+
+    """
+
+    def __init__(
+        self,
+        service_context: Optional[ServiceContext] = None,
+        storage_context: Optional[StorageContext] = None,
+        graph_query_synthesis_prompt: Optional[Prompt] = None,
+        graph_response_answer_prompt: Optional[Prompt] = None,
+        refresh_schema: bool = False,
+        verbose: bool = False,
+        **kwargs: Any,
+    ):
+        # Ensure that we have a graph store
+        assert storage_context is not None, "Must provide a storage context."
+        assert (
+            storage_context.graph_store is not None
+        ), "Must provide a graph store in the storage context."
+        self.storage_context = storage_context
+        self.graph_store = storage_context.graph_store
+
+        self.service_context = service_context or ServiceContext.from_defaults()
+
+        # Get Graph Store Type
+        self.graph_store_type = GRAPH_STORE_CLASS_TO_GRAPH_STORE_TYPE[
+            self.graph_store.__class__
+        ]
+
+        # Get Graph schema
+        self.graph_schema = self.graph_store.get_schema(refresh=refresh_schema)
+
+        # Get graph store query synthesis prompt
+        self.graph_query_synthesis_prompt = (
+            graph_query_synthesis_prompt
+            or DEFAULT_NL2GRAPH_PROMPT_MAP[self.graph_store_type]
+        )
+
+        self.graph_response_answer_prompt = (
+            graph_response_answer_prompt or DEFAULT_KG_RESPONSE_ANSWER_PROMPT
+        )
+        self.verbose = verbose
```

Comment:
Do you think some of these variables should have an `_` prefix? tbh I think almost all of them should haha except maybe the graph store?

Reply:
Sure, I agree :-D, done.